### DISCUSSION TITLE
Exercise 7: Chandy-Lamport Snapshot Protocol can build a global state that never happened.

### DISCUSSION THREAD

**Discussion by dag7dev** [sapienzastudentsnetwork/distributed-systems#32] [Exercises]
**Exercise 7:** Show that the Chandy-Lamport Snapshot Protocol can build a global state that never happened.

**Comment by AntoAndGar** [comment]
For the properties of Chandy-Lamport Snapshot Protocol a cut (or a global state, same meaning) S is (1)consistent and (2)there is a run R' that reaches S, this run is surely consistent because it respects the properties: if $e_i^1$ and $e_i^2$ are two events of the same process then $e_i^1 \rightarrow e_i^2$ or $e_i^2 \rightarrow e_i^1$  but not both are at the same time true, if $e_i$ and $e_j$ are two events of different processes and $e_i$ is a send event and $e_j$ is the corresponding receive then $e_i \rightarrow e_j$. So, we are left with the events $e_i$ and $e_j$ that are from different processes but no one is a send or receive of the same message, those types of events can be arbitrarily swapped without affecting the consistency of the global state but making a new run R != R' (or more than one) that has the same global state of S but never happened in reality.

**Comment by cornflexxx** [comment]
_Consistent Global States of Distributed Systems: Fundamental Concepts and Mechanisms_
<img width="753" height="475" alt="Screenshot 2025-11-08 at 15 12 12" src="https://github.com/user-attachments/assets/85e768a5-5b23-4a86-8ee1-43b20fa1a17b" />
<img width="823" height="389" alt="Screenshot 2025-11-08 at 15 12 07" src="https://github.com/user-attachments/assets/1625a2ca-6632-47fc-ad27-8633884c6fcb" />
